src/runtimeHelpers.ts

```diff
@@ -253,6 +298,140 @@ export function unPartitionResults<T>(
+ *
+ * Example
+ * Request args:
+ * [
+ *      { bar_id: 2, properties: ['name'], include_extra_info: true },
+ *      { bar_id: 1, properties: ['rating'], include_extra_info: false },
+ *      { bar_id: 2, properties: ['rating'], include_extra_info: true },
+ * ]
+ *
+ * ```js
+ * unPartitionResultsByBatchKeyPartition(
+ *   newKey = 'bar_id',
+ *   propertyBatchKey =  'properties',
+ *   responseKey = 'id'
+ *   batchKeyPartition = [ [2, 2], [1] ],
+ *   propertyBatchKeyPartion = [ [['name'], ['rating']], [['rating']] ],
+ *   requestGroups = [ [0, 2], [1] ],
+ *   resultGroups = [
+ *      [ { id: 2, name: 'Burger King', rating: 3 } ],
+ *      [ { id: 1, name: 'In N Out', rating: 4 } ]
+ *   ],
+ * )
+ * ```
+ *
+ * Returns:
+ * ```
+ * [
+ *   { id: 2, name: 'Burger King' }，
+ *   { id: 1, rating: 4 }，
+ *   { id: 2, rating: 3 }，
+ * ]
+ */
+export function unPartitionResultsByBatchKeyPartition<T extends Record<string, any>>(
+    newKey: string,
+    propertyBatchKey: string,
+    responseKey: string,
+    batchKeyPartition: ReadonlyArray<ReadonlyArray<any>>,
+    propertyBatchKeyPartion: ReadonlyArray<ReadonlyArray<any>>,
+    /** Should be a nested array of IDs, as generated by partitionItems */
+    requestGroups: ReadonlyArray<ReadonlyArray<number>>,
+    /** The results back from the service, in the same shape as groups */
+    resultGroups: ReadonlyArray<ReadonlyArray<T | CaughtResourceError>>,
+): ReadonlyArray<T | Error> {
+    /**
+     * e.g. with our inputs, produce:
+     * ```js
+     * [
+     *   [
+     *      { order: 0, result: { id: 2, name: 'Burger King' },
+     *      { order: 2, result: { id: 2, rating: 3 } },
+     *   ],
+     *   [
+     *      { order: 1, result: { id: 1, rating: 4 } },
+     *   ]
+     * ]
+     * ```
+     */
+    const zippedGroups: ReadonlyArray<ReadonlyArray<{ order: number; result: T | Error }>> = requestGroups.map(
+        (ids, i) => {
+            return ids.map((id, j) => {
+                let result = null;
+                for (const resultElement of Object.values(resultGroups)[i]) {
+                    // There's error in the result we should return
+                    if (resultElement instanceof CaughtResourceError) {
+                        result = resultElement;
+                        break;
+                    }
+
+                    // Find the response that matches the requested batchKey
+                    if (resultElement[responseKey] === batchKeyPartition[i][j]) {
+                        // Only responseKey and the requested properties will be in the final result.
+                        result = _.pick(resultElement, [responseKey, ...propertyBatchKeyPartion[i][j]]);
+                        break;
+                    }
+                }
+                // If requested property doesn't exist in resultElement, we should throw BatchItemNotFoundError error.
+                if (result === null) {
```

Comment:
```suggestion
                }

                // If requested property doesn't exist in resultElement, we should throw BatchItemNotFoundError error.
                if (result === null) {
```

nit: spacing between stanzas